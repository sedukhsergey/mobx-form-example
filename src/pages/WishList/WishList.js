import React, { useEffect } from 'react';
import { inject, observer } from "mobx-react";
import styles from './styles.module.css';
import SantaImage from '../../assets/santa.jpeg';
import { WishListView } from "../../modules";


const WishList = inject('mobxStore')(observer(props => {
  const {
    mobxStore: {
      wishListStore: {
        items,
        add,
        totalPrice,
        fetchItems
      }
    },
  } = props;
  console.log('items',items)
  useEffect(() => {
      if (!items.length) {
        fetchItems()
      }
  },[]);
  return (
    <div className={styles.container}>
      <header className={styles.hero}>
        <h1>Wish List</h1>
        <img src={SantaImage} alt="santa"/>
      </header>
      <WishListView wishList={items}/>
    </div>
  )
}));

export default WishList;
