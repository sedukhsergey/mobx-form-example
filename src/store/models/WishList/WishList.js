import { types, destroy } from 'mobx-state-tree';
import { WishListItem } from "./index";


const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), []),
  })
  .actions(self => ({
    add(item) {
      self.items.push(item);
    },
  }))
  .views(self => ({
    get totalPrice() {
      return self.items.reduce((acc, curr) => acc + curr.price,0)
    }
  }))


export default WishList
