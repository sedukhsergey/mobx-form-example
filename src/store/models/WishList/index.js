export { default as WishList } from './WishList';
export { default as WishListItem } from './WishListItem';
