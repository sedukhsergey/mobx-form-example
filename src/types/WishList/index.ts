export type WishListItem = {
  name: string,
  price: number,
  image: string,
  changeName: (item: string) => void,
  changePrice: (item: number) => void,
  changeImage: (item: string) => void,
  remove: () => void,
}
