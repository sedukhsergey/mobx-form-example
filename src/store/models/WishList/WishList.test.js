import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { reaction } from "mobx";
import WishList from "./WishList";

const data = {
  "name": "Default name",
  "price": 34.23,
  "image": "https://via.placeholder.com/150"
}

it('WishList test', () => {
  const list = WishList.create();
  const states = [];
  onSnapshot(list, snapshot => {
    states.push(snapshot);
  })
  list.add(data)
  expect(list.items.length).toBe(1);
  list.items[0].changeName('Another name');
  expect(list.items[0].name).toBe('Another name');

  expect(getSnapshot(list))
  //   .toEqual({
  //   items: [
  //     {
  //       name: 'Another name',
  //       price: 34.23,
  //       image: "https://via.placeholder.com/150"
  //     }
  //   ]
  // })
    .toMatchSnapshot()
  expect(states).toMatchSnapshot();
});

it('WishList test - 2', () => {
  const list = WishList.create();
  const patches = [];
  onPatch(list, patch => {
    patches.push(patch);
  })
  list.add(data)
  expect(list.items.length).toBe(1);
  list.items[0].changeName('Another name');
  expect(list.items[0].name).toBe('Another name');

  expect(getSnapshot(list))
  //   .toEqual({
  //   items: [
  //     {
  //       name: 'Another name',
  //       price: 34.23,
  //       image: "https://via.placeholder.com/150"
  //     }
  //   ]
  // })
    .toMatchSnapshot()
  expect(patches).toMatchSnapshot();
});

it('can calculate totalPrice', () => {
  const list = WishList.create();
  list.add({
    "name": "First Item",
    "price": 42,
    "image": "https://via.placeholder.com/150"
  })
  list.add({
    "name": "Second Item",
    "price": 52,
    "image": "https://via.placeholder.com/150"
  })
  expect(list.totalPrice).toBe(94);
  let changed = 0;
  reaction(() => list.totalPrice, () => changed++);
  expect(changed).toBe(0);
  console.log('totalPrice',list.totalPrice)
  list.items[0].changeName('test');
  expect(changed).toBe(0);
  list.items[0].changePrice(23);
  expect(changed).toBe(1);
})

