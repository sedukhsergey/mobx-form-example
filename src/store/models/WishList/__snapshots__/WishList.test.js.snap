// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishList test - 2 1`] = `
Object {
  "items": Array [
    Object {
      "image": "https://via.placeholder.com/150",
      "name": "Another name",
      "price": 34.23,
    },
  ],
}
`;

exports[`WishList test - 2 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "https://via.placeholder.com/150",
      "name": "Default name",
      "price": 34.23,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Another name",
  },
]
`;

exports[`WishList test 1`] = `
Object {
  "items": Array [
    Object {
      "image": "https://via.placeholder.com/150",
      "name": "Another name",
      "price": 34.23,
    },
  ],
}
`;

exports[`WishList test 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "https://via.placeholder.com/150",
        "name": "Default name",
        "price": 34.23,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "https://via.placeholder.com/150",
        "name": "Another name",
        "price": 34.23,
      },
    ],
  },
]
`;
